src/cursor.rs

```diff
@@ -0,0 +1,150 @@
+use core::fmt;
+use std::{error::Error, sync::Arc};
+
+use crate::{icon::PIXEL_SIZE, platform_impl::PlatformCustomCursor};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CustomCursor {
+    pub(crate) inner: Arc<PlatformCustomCursor>,
+}
+
+impl CustomCursor {
+    pub fn from_rgba(
+        rgba: Vec<u8>,
+        width: u32,
+        height: u32,
+        hotspot_x: u32,
+        hotspot_y: u32,
+    ) -> Result<Self, BadImage> {
+        Ok(Self {
+            inner: PlatformCustomCursor::from_rgba(rgba, width, height, hotspot_x, hotspot_y)?
+                .into(),
+        })
+    }
+}
+
+/// Platforms that don't support cursors will export this as `PlatformCustomCursor`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct NoCustomCursor;
+
+#[allow(dead_code)]
+impl NoCustomCursor {
+    pub fn from_rgba(
+        rgba: Vec<u8>,
+        width: u32,
+        height: u32,
+        hotspot_x: u32,
+        hotspot_y: u32,
+    ) -> Result<Self, BadImage> {
+        CursorImage::from_rgba(rgba, width, height, hotspot_x, hotspot_y)?;
+        Ok(Self)
+    }
+}
+
+#[derive(Debug)]
+/// An error produced when using [`Icon::from_rgba`] with invalid arguments.
+pub enum BadImage {
+    /// Produced when the length of the `rgba` argument isn't divisible by 4, thus `rgba` can't be
+    /// safely interpreted as 32bpp RGBA pixels.
+    ByteCountNotDivisibleBy4 { byte_count: usize },
```

Comment:
Consideration (that you may have already talked about, if so, I would like to see the reasoning be stated in a code comment):

`softbuffer` [uses](https://docs.rs/softbuffer/latest/softbuffer/struct.Surface.html#method.fetch) `u32` instead of `u8` in its buffer, exactly so that errors like this are entirely avoided.

Reply:
It doesn't have reasoning other than that I copied it from `src/icon.rs` [link](https://github.com/rust-windowing/winit/blob/7bed5eecfdcbde16e5619fd137f0229e8e7e8ed4/src/icon.rs#L20C26-L20C26).

I'm fine with using u32, but what about endianess? Currently it is using big endian RGBA pixels, which work directly only in web and macos. Windows, X11 and Wayland want little endian ARGB pixels and need extra processing currently.

Reply:
@eero-lehtinen endianess is explicitly stated in the underlying pixel formats. On Wayland it's always little endian, regardless of used arch, maybe the same on X11.

I'd suggest to follow softbuffer approach to buffers here, since it's known to work and using u32 avoids a lot of errors.

The users need to decode images anyway, and they need to ensure that they have alpha properly in them, like they can't just pass data as is.

Reply:
I'm not particularly convinced that `softbuffer` has done the right thing here. Even though endianess seems to be defined for Wayland it isn't clear in other backends/OS's (AFAIK). I wrote up some of my thoughts here: https://github.com/rust-windowing/softbuffer/issues/109.

I'm strongly in favor of leaving it as is, as the same API is already used in `Icon`. We can change the API in a follow-up issue/PR where we can have that discussion and change both and not leave them differently.

Reply:
I would like to leave it as is. I already made it and it already works :D. 

The change would be annoying, because x11 is the only one that accepts u32, so all others would need to be manually converted to u8. This change would remove this single user facing error but not much else, I think. But I'll leave this up to you to choose.